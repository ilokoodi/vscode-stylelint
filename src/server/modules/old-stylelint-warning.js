'use strict';

const path = require('path');
const fs = require('fs/promises');
const semver = require('semver');
const { getWorkspaceFolder } = require('../../utils/documents');
const { findPackageRoot } = require('../../utils/packages');

/**
 * @implements {LanguageServerModule}
 */
class OldStylelintWarningModule {
	static id = 'old-stylelint-warning';

	/**
	 * The language server context.
	 * @type {LanguageServerContext}
	 */
	#context;

	/**
	 * The logger to use, if any.
	 * @type {winston.Logger | undefined}
	 */
	#logger;

	/**
	 * Set of workspaces for which Stylelint's version has already been checked.
	 * @type {Set<string>}
	 */
	#checkedWorkspaces = new Set();

	/**
	 * Whether or not to provide the URL to the migration guide.
	 * @type {boolean}
	 */
	#openMigrationGuide = false;

	/**
	 * @param {LanguageServerModuleConstructorParameters} params
	 */
	constructor({ context, logger }) {
		this.#context = context;
		this.#logger = logger;
	}

	/**
	 * @param {lsp.InitializeParams} params
	 * @returns {void}
	 */
	onInitialize({ capabilities }) {
		this.#openMigrationGuide = capabilities.window?.showDocument?.support ?? false;
	}

	/**
	 * @param {lsp.TextDocument} document
	 * @returns {Promise<string | undefined>}
	 */
	async #getStylelintVersion(document) {
		const result = await this.#context.resolveStylelint(document);

		if (!result) {
			this.#logger?.debug('Stylelint not found', {
				uri: document.uri,
			});

			return undefined;
		}

		const packageDir = await findPackageRoot(result.resolvedPath);

		if (!packageDir) {
			this.#logger?.debug('Stylelint package root not found', {
				uri: document.uri,
			});

			return undefined;
		}

		const manifestPath = path.join(packageDir, 'package.json');

		try {
			const rawManifest = await fs.readFile(manifestPath, 'utf8');

			const manifest = JSON.parse(rawManifest);

			return manifest.version;
		} catch (error) {
			this.#logger?.debug('Stylelint package manifest could not be read', {
				uri: document.uri,
				manifestPath,
				error,
			});

			return undefined;
		}
	}

	/**
	 * @param {lsp.TextDocument} document
	 * @returns {Promise<string | undefined>}
	 */
	async #check(document) {
		if (!this.#context.options.validate.includes(document.languageId)) {
			this.#logger?.debug('Document should not be validated, ignoring', {
				uri: document.uri,
				language: document.languageId,
			});

			return undefined;
		}

		const workspaceFolder = await getWorkspaceFolder(this.#context.connection, document);

		if (!workspaceFolder) {
			this.#logger?.debug('Document not part of a workspace, ignoring', {
				uri: document.uri,
			});

			return undefined;
		}

		if (this.#checkedWorkspaces.has(workspaceFolder)) {
			this.#logger?.debug('Document has already been checked, ignoring', {
				uri: document.uri,
			});

			return undefined;
		}

		this.#checkedWorkspaces.add(workspaceFolder);

		const stylelintVersion = await this.#getStylelintVersion(document);

		try {
			return stylelintVersion && semver.lt(stylelintVersion, '14.0.0')
				? stylelintVersion
				: undefined;
		} catch (error) {
			this.#logger?.debug('Stylelint version could not be parsed', {
				uri: document.uri,
				version: stylelintVersion,
				error,
			});

			return undefined;
		}
	}

	/**
	 * @returns {void}
	 */
	onDidRegisterHandlers() {
		this.#logger?.debug('Registering onDidOpen handler');

		this.#context.documents.onDidOpen(async ({ document }) => {
			const stylelintVersion = await this.#check(document);

			if (!stylelintVersion) {
				return;
			}

			this.#logger?.warn(`Found unsupported version of Stylelint: ${stylelintVersion}`);

			const message = `Stylelint version ${stylelintVersion} is no longer supported â€” you may encounter unexpected behavior. Please upgrade to version 14.0.0 or newer. See the migration guide for more information.`;

			if (!this.#openMigrationGuide) {
				this.#context.connection.window.showWarningMessage(message);

				return;
			}

			const warningResponse = await this.#context.connection.window.showWarningMessage(message, {
				title: 'Open migration guide',
			});

			if (warningResponse?.title === 'Open migration guide') {
				// Open URL in browser
				const showURIResponse = await this.#context.connection.window.showDocument({
					uri: 'https://github.com/stylelint/vscode-stylelint#migrating-from-vscode-stylelint-0xstylelint-13x',
					external: true,
				});

				if (!showURIResponse.success) {
					this.#logger?.warn('Failed to open migration guide');
				}
			}
		});

		this.#logger?.debug('onDidOpen handler registered');
	}
}

module.exports = {
	OldStylelintWarningModule,
};
