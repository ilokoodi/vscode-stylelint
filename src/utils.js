const fs = require('fs').promises;
const path = require('path');

/**
 * Walks up the file tree from the current directory until it finds a
 * directory containing `package.json`. Resolves to `undefined` if no such
 * directory is found.
 * @param {string} directory The directory to start from.
 * @returns {Promise<string | undefined>}
 */
async function resolvePackageDirectory(directory) {
	let currentDirectory = directory;

	// eslint-disable-next-line no-constant-condition
	while (true) {
		const manifestPath = path.join(currentDirectory, 'package.json');

		try {
			await fs.access(manifestPath);

			return currentDirectory;
		} catch (error) {
			if (/** @type {{code?: string}} */ (error).code === 'ENOENT') {
				const parent = path.dirname(currentDirectory);

				if (!path.relative(parent, currentDirectory)) {
					return undefined;
				}

				currentDirectory = parent;
			} else {
				throw error;
			}
		}
	}
}

module.exports = {
	resolvePackageDirectory,
};
