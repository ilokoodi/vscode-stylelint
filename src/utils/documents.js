'use strict';

const pathIsInside = require('path-is-inside');
const { URI } = require('vscode-uri');

/**
 * @param {lsp.TextDocument} document
 * @param {lsp.Connection} [connection]
 * @returns {Promise<string | undefined>}
 */
async function getWorkspaceFolder(document, connection) {
	const documentPath = URI.parse(document.uri).fsPath;

	if (documentPath) {
		const workspaceFolders = await connection?.workspace.getWorkspaceFolders();

		if (workspaceFolders) {
			for (const { uri } of workspaceFolders) {
				const workspacePath = URI.parse(uri).fsPath;

				if (pathIsInside(documentPath, workspacePath)) {
					return workspacePath;
				}
			}
		}
	}

	return undefined;
}

module.exports = {
	getWorkspaceFolder,
};
