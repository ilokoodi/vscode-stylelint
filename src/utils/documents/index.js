'use strict';

module.exports = {
	...require('./create-text-edits'),
	...require('./get-disable-type'),
	...require('./get-fixes'),
	...require('./get-workspace-folder'),
};
