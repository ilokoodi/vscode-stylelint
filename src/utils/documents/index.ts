export * from './create-text-edits';
export * from './get-disable-type';
export * from './get-fixes';
export * from './get-workspace-folder';
