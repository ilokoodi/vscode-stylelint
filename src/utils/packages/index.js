'use strict';

module.exports = {
	...require('./global-path-resolver'),
	...require('./find-package-root'),
};
