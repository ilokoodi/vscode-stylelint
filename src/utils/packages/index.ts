export * from './find-package-root';
export * from './global-path-resolver';
export * from './stylelint-resolver';
export * from './types';
