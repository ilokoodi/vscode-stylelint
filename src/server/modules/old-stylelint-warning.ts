import path from 'path';
import fs from 'fs/promises';
import semver from 'semver';
import pathIsInside from 'path-is-inside';
import type { TextDocument } from 'vscode-languageserver-textdocument';
import type LSP from 'vscode-languageserver-protocol';
import { getWorkspaceFolder } from '../../utils/documents/index';
import { findPackageRoot } from '../../utils/packages/index';
import type {
	LanguageServerContext,
	LanguageServerModuleConstructorParameters,
	LanguageServerModule,
} from '../types';
import type winston from 'winston';

export class OldStylelintWarningModule implements LanguageServerModule {
	static id = 'old-stylelint-warning';

	/**
	 * The language server context.
	 */
	#context: LanguageServerContext;

	/**
	 * The logger to use, if any.
	 */
	#logger: winston.Logger | undefined;

	/**
	 * Set of workspaces for which Stylelint's version has already been checked.
	 */
	#checkedWorkspaces = new Set<string>();

	/**
	 * Whether or not to provide the URL to the migration guide.
	 */
	#openMigrationGuide = false;

	/**
	 * Disposables for handlers.
	 */
	#disposables: LSP.Disposable[] = [];

	constructor({ context, logger }: LanguageServerModuleConstructorParameters) {
		this.#context = context;
		this.#logger = logger;
	}

	dispose(): void {
		this.#disposables.forEach((disposable) => disposable.dispose());
		this.#disposables.length = 0;
	}

	onInitialize({ capabilities }: LSP.InitializeParams): void {
		this.#openMigrationGuide = capabilities.window?.showDocument?.support ?? false;
	}

	async #getStylelintVersion(
		document: TextDocument,
		workspaceFolder: string,
	): Promise<string | undefined> {
		const result = await this.#context.resolveStylelint(document);

		if (!result) {
			this.#logger?.debug('Stylelint not found', {
				uri: document.uri,
			});

			return undefined;
		}

		const packageDir = await findPackageRoot(result.resolvedPath);

		if (!packageDir) {
			this.#logger?.debug('Stylelint package root not found', {
				uri: document.uri,
			});

			return undefined;
		}

		if (!pathIsInside(packageDir, workspaceFolder)) {
			this.#logger?.debug('Stylelint package root is not inside the workspace', {
				uri: document.uri,
			});

			return undefined;
		}

		const manifestPath = path.join(packageDir, 'package.json');

		try {
			const rawManifest = await fs.readFile(manifestPath, 'utf8');

			const manifest = JSON.parse(rawManifest) as { version?: string };

			return manifest.version;
		} catch (error) {
			this.#logger?.debug('Stylelint package manifest could not be read', {
				uri: document.uri,
				manifestPath,
				error,
			});

			return undefined;
		}
	}

	async #check(document: TextDocument): Promise<string | undefined> {
		const workspaceFolder = await getWorkspaceFolder(this.#context.connection, document);

		if (!workspaceFolder) {
			this.#logger?.debug('Document not part of a workspace, ignoring', {
				uri: document.uri,
			});

			return undefined;
		}

		if (this.#checkedWorkspaces.has(workspaceFolder)) {
			this.#logger?.debug('Document has already been checked, ignoring', {
				uri: document.uri,
			});

			return undefined;
		}

		this.#checkedWorkspaces.add(workspaceFolder);

		const stylelintVersion = await this.#getStylelintVersion(document, workspaceFolder);

		if (!stylelintVersion) {
			return undefined;
		}

		try {
			const coerced = semver.coerce(stylelintVersion);

			if (!coerced) {
				throw new Error(`Could not coerce version "${stylelintVersion}"`);
			}

			return semver.lt(coerced, '14.0.0') ? stylelintVersion : undefined;
		} catch (error) {
			this.#logger?.debug('Stylelint version could not be parsed', {
				uri: document.uri,
				version: stylelintVersion,
				error,
			});

			return undefined;
		}
	}

	onDidRegisterHandlers(): void {
		this.#logger?.debug('Registering onDidOpen handler');

		this.#disposables.push(
			this.#context.documents.onDidOpen(async ({ document }) => {
				const stylelintVersion = await this.#check(document);

				if (!stylelintVersion) {
					return;
				}

				this.#logger?.warn(`Found unsupported version of Stylelint: ${stylelintVersion}`);

				const message = `Stylelint version ${stylelintVersion} is no longer supported. While it may continue to work for a while, you may encounter unexpected behavior. Please upgrade to version 14.0.0 or newer. See the migration guide for more information.`;

				if (!this.#openMigrationGuide) {
					this.#context.connection.window.showWarningMessage(message);

					return;
				}

				const warningResponse = await this.#context.connection.window.showWarningMessage(message, {
					title: 'Open migration guide',
				});

				if (warningResponse?.title === 'Open migration guide') {
					// Open URL in browser
					const showURIResponse = await this.#context.connection.window.showDocument({
						uri: 'https://github.com/stylelint/vscode-stylelint#migrating-from-vscode-stylelint-0xstylelint-13x',
						external: true,
					});

					if (!showURIResponse.success) {
						this.#logger?.warn('Failed to open migration guide');
					}
				}
			}),
		);

		this.#logger?.debug('onDidOpen handler registered');
	}
}
