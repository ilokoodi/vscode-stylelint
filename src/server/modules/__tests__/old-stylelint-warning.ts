jest.mock('../../../utils/documents');
jest.mock('../../../utils/packages');
jest.mock('fs/promises');
jest.mock('path');

import fs from 'fs/promises';
import path from 'path';
import type LSP from 'vscode-languageserver-protocol';
import type winston from 'winston';
import { getWorkspaceFolder } from '../../../utils/documents';
import { findPackageRoot } from '../../../utils/packages';
import { OldStylelintWarningModule } from '../old-stylelint-warning';
import type { LanguageServerModuleConstructorParameters } from '../../types';

const mockedFS = fs as tests.mocks.FSPromisesModule;
const mockedPath = path as tests.mocks.PathModule;
const mockedGetWorkspaceFolder = getWorkspaceFolder as jest.MockedFunction<
	typeof getWorkspaceFolder
>;
const mockedFindPackageRoot = findPackageRoot as jest.MockedFunction<typeof findPackageRoot>;

const mockContext = {
	connection: {
		window: {
			showWarningMessage: jest.fn(),
			showDocument: jest.fn(),
		},
	},
	documents: { onDidOpen: jest.fn() },
	options: { validate: [] as string[] },
	displayError: jest.fn(),
	resolveStylelint: jest.fn(),
};

const mockLogger = {
	debug: jest.fn(),
	warn: jest.fn(),
	error: jest.fn(),
} as unknown as jest.Mocked<winston.Logger>;

const getParams = (passLogger = false) =>
	({
		context: mockContext,
		logger: passLogger ? mockLogger : undefined,
	} as unknown as LanguageServerModuleConstructorParameters);

describe('OldStylelintWarningModule', () => {
	beforeEach(() => {
		mockedPath.__mockPlatform('posix');
		mockContext.options.validate = [];
		jest.clearAllMocks();
	});

	test('should be constructable', () => {
		expect(() => new OldStylelintWarningModule(getParams())).not.toThrow();
	});

	test('onDidRegisterHandlers should register an onDidOpen handler', () => {
		const module = new OldStylelintWarningModule(getParams());

		module.onDidRegisterHandlers();

		expect(mockContext.documents.onDidOpen).toHaveBeenCalledTimes(1);
		expect(mockContext.documents.onDidOpen).toHaveBeenCalledWith(expect.any(Function));
	});

	test('if document language ID is not in options, should not warn', async () => {
		mockContext.options.validate = ['baz'];

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({ capabilities: {} } as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).not.toHaveBeenCalled();
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Document should not be validated, ignoring',
			{ uri: 'foo', language: 'bar' },
		);
	});

	test('if document is not part of a workspace, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue(undefined);
		mockContext.options.validate = ['bar'];

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({ capabilities: {} } as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).not.toHaveBeenCalled();
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Document not part of a workspace, ignoring',
			{ uri: 'foo' },
		);
	});

	test('if document has already been checked, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockContext.options.validate = ['bar'];

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({ capabilities: {} } as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Document has already been checked, ignoring',
			{ uri: 'foo' },
		);
	});

	test('if Stylelint package root cannot be determined, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue(undefined);
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {},
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.options.validate = ['bar'];

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({ capabilities: {} } as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Stylelint package root not found', {
			uri: 'foo',
		});
	});

	test('if Stylelint package manifest cannot be read, should not warn', async () => {
		const error = new Error('foo');

		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {},
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.options.validate = ['bar'];
		mockedFS.readFile.mockRejectedValue(error);

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({ capabilities: {} } as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Stylelint package manifest could not be read',
			{ uri: 'foo', manifestPath: '/path/node_modules/stylelint/package.json', error },
		);
	});

	test('if Stylelint package manifest is malformed, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {},
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.options.validate = ['bar'];
		mockedFS.readFile.mockResolvedValue('{');

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({ capabilities: {} } as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Stylelint package manifest could not be read',
			{
				uri: 'foo',
				manifestPath: '/path/node_modules/stylelint/package.json',
				error: expect.any(Error),
			},
		);
	});

	test('if Stylelint package manifest does not contain a version, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {},
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.options.validate = ['bar'];
		mockedFS.readFile.mockResolvedValue('{}');

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({ capabilities: {} } as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
	});

	test('if Stylelint version cannot be parsed, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {},
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.options.validate = ['bar'];
		mockedFS.readFile.mockResolvedValue('{"version": "foo"}');

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({ capabilities: {} } as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Stylelint version could not be parsed', {
			uri: 'foo',
			version: 'foo',
			error: expect.any(Error),
		});
	});

	test('if Stylelint version is 14.x or greater, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {},
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.options.validate = ['bar'];
		mockedFS.readFile.mockResolvedValue('{"version": "14.0.0"}');

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({ capabilities: {} } as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
	});

	test('if Stylelint version is 14.x with a label, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {},
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.options.validate = ['bar'];
		mockedFS.readFile.mockResolvedValue('{"version": "14.0.0-sdk"}');

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({ capabilities: {} } as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
	});

	test('without openDocument support, if Stylelint version is less than 14.x, should warn and provide link to migration guide', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {},
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.options.validate = ['bar'];
		mockedFS.readFile.mockResolvedValue('{"version": "13.0.0"}');

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({ capabilities: {} } as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage.mock.calls[0]).toMatchSnapshot();
		expect(mockContext.connection.window.showDocument).not.toHaveBeenCalled();
	});

	test("with openDocument support, if Stylelint version is less than 14.x and user doesn't click button, should warn but not open URL", async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {},
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.options.validate = ['bar'];
		mockContext.connection.window.showWarningMessage.mockResolvedValue(undefined);
		mockedFS.readFile.mockResolvedValue('{"version": "13.0.0"}');

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({
			capabilities: {
				window: {
					showDocument: { support: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showDocument).not.toHaveBeenCalled();
	});

	test('with openDocument support, if Stylelint version is less than 14.x and user clicks button, should warn and open URL', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {},
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.options.validate = ['bar'];
		mockContext.connection.window.showWarningMessage.mockResolvedValue({
			title: 'Open migration guide',
		});
		mockContext.connection.window.showDocument.mockResolvedValue({
			success: true,
		});
		mockedFS.readFile.mockResolvedValue('{"version": "13.0.0"}');

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({
			capabilities: {
				window: {
					showDocument: { support: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showDocument).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showDocument.mock.calls[0]).toMatchSnapshot();
		expect(mockLogger.warn).not.toHaveBeenCalledWith('Failed to open migration guide');
	});

	test('with openDocument support, if Stylelint version is less than 14.x and user clicks button, but fails to open URL, should warn and log', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {},
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.options.validate = ['bar'];
		mockContext.connection.window.showWarningMessage.mockResolvedValue({
			title: 'Open migration guide',
		});
		mockContext.connection.window.showDocument.mockResolvedValue({
			success: false,
		});
		mockedFS.readFile.mockResolvedValue('{"version": "13.0.0"}');

		const module = new OldStylelintWarningModule(getParams(true));

		module.onInitialize({
			capabilities: {
				window: {
					showDocument: { support: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: { uri: 'foo', languageId: 'bar' },
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showDocument).toHaveBeenCalledTimes(1);
		expect(mockLogger.warn).toHaveBeenCalledWith('Failed to open migration guide');
	});
});
