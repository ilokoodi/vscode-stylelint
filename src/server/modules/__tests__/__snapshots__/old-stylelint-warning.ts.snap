// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OldStylelintWarningModule with openDocument support, if Stylelint version is less than 14.x and user clicks button, should warn and open URL 1`] = `
Array [
  Object {
    "external": true,
    "uri": "https://github.com/stylelint/vscode-stylelint#migrating-from-vscode-stylelint-0xstylelint-13x",
  },
]
`;

exports[`OldStylelintWarningModule without openDocument support, if Stylelint version is less than 14.x, should warn and provide link to migration guide 1`] = `
Array [
  "Stylelint version 13.0.0 is no longer supported â€” you may encounter unexpected behavior. Please upgrade to version 14.0.0 or newer. See the migration guide for more information.",
]
`;
